### PR TITLE
Add theming support for hovered ItemList items

### LINKED ISSUE #6478: Implement hovered item style in ItemList
### Describe the project you are working on

A data editing tool, more like a traditional desktop application. There are list views and icon grid views, taking advantage of the ItemList widget.

I would like to have the list rows and icon grid items highlighted when hovered by the mouse.

This change is also proposed as an improvement to Godot itself.
Same principle can apply to Tree controls.

Examples implemented on the master branch at time of writing:
(mouse pointer not shown as my capture software does not handle it, but you can guess the moving highlight corresponds to the hovering)

list1

list2

### Describe the problem or limitation you are having in your project

The ItemList widget does not have a separate style for the hovered item.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

The ItemList control is modified to support a hover style, applied to the currently hovered item.
The default & editor themes are updated in consequence.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

Already implemented on ItemList: https://github.com/godotengine/godot/pull/74739

The control is modified to track changes in hovered items at each mouse movement. A new StyleBox and font color are used to style the item being hovered.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

Extracting the whole ItemList control for reuse is inconvenient.

One approximation would be to throw a bunch of buttons in a scroll container + VBox and style those, but a big part of the list selection behaviour would need to be reimplemented. In addition, it is less efficient.

### Is there a reason why this should be core and not an add-on in the asset library?

This impacts a core widget.

### PR DESCRIPTION
**Pull request by davthedev**
Addition of a distinct styling for the hovered item in ItemList controls.

Comes with the corresponding changes in default and editor themes.

Feature proposal with visuals: https://github.com/godotengine/godot-proposals/issues/6478

Can be backported to 3.x with adaptations to the branch themes.

_Production edit: Closes https://github.com/godotengine/godot-proposals/issues/6478_

### CODE REVIEW

File: doc/classes/ItemList.xml [hunk 1]
```diff
@@ -479,6 +482,9 @@
 		<theme_item name="focus" data_type="style" type="StyleBox">
 			The focused style for the [ItemList], drawn on top of the background, but below everything else.
 		</theme_item>
+		<theme_item name="hovered" data_type="style" type="StyleBox">
+			[StyleBox] for the hovered, not yet selected items.
```
Comment (KoBeWi) [inline doc/classes/ItemList.xml:486]: Why is this description different from `font_hovered_color`, if their logic is the same?

File: scene/gui/item_list.cpp [hunk 2]
```diff
@@ -1436,6 +1443,11 @@ void ItemList::_scroll_changed(double) {
 	queue_redraw();
 }
 
+void ItemList::_mouse_exited() {
+	hovered = -1;
```
Comment (KoBeWi) [inline scene/gui/item_list.cpp:1447]: You could add a condition `if hovered > -1`.

File: doc/classes/ItemList.xml (hunk 1 above)
Comment (davthedev) [reply doc/classes/ItemList.xml:486]: Ajusted, the logic is the same

File: scene/gui/item_list.cpp [hunk 3]
```diff
@@ -646,6 +646,15 @@ void ItemList::gui_input(const Ref<InputEvent> &p_event) {
 		return;
 	}
 
+	if (mm.is_valid()) {
+		int closest = get_item_at_position(mm->get_position(), true);
+		int old_hovered = hovered;
```
[the remark this answers is not in the document]
Comment (davthedev) [reply scene/gui/item_list.cpp]: Fixed

File: doc/classes/ItemList.xml (hunk 1 above)
Comment (KoBeWi) [inline doc/classes/ItemList.xml:486]:
```suggestion
			[StyleBox] for the hovered, but not selected items.
```
I think "yet" is odd here. Same above.
Although maybe wording can be better than what I suggested.

**Review by KoBeWi — APPROVED** [review]

File: scene/resources/default_theme/default_theme.cpp [hunk 4]
```diff
@@ -803,9 +803,11 @@ void fill_default_theme(Ref<Theme> &theme, const Ref<Font> &default_font, const
 	theme->set_font_size("font_size", "ItemList", -1);
 
 	theme->set_color("font_color", "ItemList", control_font_lower_color);
+	theme->set_color("font_hovered_color", "ItemList", control_font_hover_color);
 	theme->set_color("font_selected_color", "ItemList", control_font_pressed_color);
 	theme->set_color("font_outline_color", "ItemList", Color(1, 1, 1));
 	theme->set_color("guide_color", "ItemList", Color(0.7, 0.7, 0.7, 0.25));
+	theme->set_stylebox("hovered", "ItemList", make_flat_stylebox(Color(1, 1, 1, 0.07)));
```
Comment (YuriSizov) [inline scene/resources/default_theme/default_theme.cpp:810]: Default theme should not be changed and should provide values consistent with the current look, so existing projects keep their appearance.

File: scene/gui/item_list.cpp [hunk 5]
```diff
@@ -1853,6 +1866,7 @@ ItemList::ItemList() {
 	add_child(scroll_bar, false, INTERNAL_MODE_FRONT);
 
 	scroll_bar->connect("value_changed", callable_mp(this, &ItemList::_scroll_changed));
+	connect("mouse_exited", callable_mp(this, &ItemList::_mouse_exited));
```
Comment (YuriSizov) [inline scene/gui/item_list.cpp:1869]:
```suggestion
	scroll_bar->connect("value_changed", callable_mp(this, &ItemList::_scroll_changed));

	connect("mouse_exited", callable_mp(this, &ItemList::_mouse_exited));
```

This new line is unrelated to the scrollbar, so it shouldn't be visually grouped with it. And scrollbar's own connect method can be moved next to its related code above.

File: scene/gui/item_list.cpp [hunk 6]
```diff
@@ -1204,7 +1210,7 @@ void ItemList::_notification(int p_what) {
 						max_len = size2.x;
 					}
 
-					Color txt_modulate = items[i].selected ? theme_cache.font_selected_color : (items[i].custom_fg != Color() ? items[i].custom_fg : theme_cache.font_color);
+					Color txt_modulate = items[i].selected ? theme_cache.font_selected_color : (hovered == i ? theme_cache.font_hovered_color : (items[i].custom_fg != Color() ? items[i].custom_fg : theme_cache.font_color));
```
Comment (YuriSizov) [inline scene/gui/item_list.cpp:1213]: I think we're at a point where this line needs to be unrolled into a proper if-else.

**Review by YuriSizov — CHANGES_REQUESTED** [review]
Left a few notes. Looks good to me otherwise.

File: scene/resources/default_theme/default_theme.cpp (hunk 4 above)
Comment (YuriSizov) [reply scene/resources/default_theme/default_theme.cpp:810]: We are in a stable phase right now and must preserve compatibility whenever possible. There are no expectations that Godot 3 projects will work and look the same in Godot 4. There is such expectation between 4.0 and 4.1.

There is no reason why this should be enabled by default for everyone when this wasn't the case before. Those who need this for their project will update their theme accordingly.

Comment (KoBeWi) [reply scene/resources/default_theme/default_theme.cpp:810]: I think people who use the default theme in their project don't expect that it will never change at all. And this isn't some big change, it's just a highlight for list items on mouse hover. Most people will appreciate it or won't mind at all.

Comment (YuriSizov) [reply scene/resources/default_theme/default_theme.cpp:810]: This affects everyone, not just people who never change their theme from the default. Since this adds new properties, they will not be overridden in user themes.

Comment (Calinou) [reply scene/resources/default_theme/default_theme.cpp:810]: See also https://github.com/godotengine/godot/pull/29268.

I would really prefer this to be enabled by default in projects. Otherwise, it's hardly going to see any use at all.

Comment (YuriSizov) [reply scene/resources/default_theme/default_theme.cpp:810]: We don't really plan any special compatibility breakage for 4.1 which would not be possible for any future 4.x. The same rules will apply going forward and starting now.

It is indeed a small change, so if you insist, alright then, let's keep it.

Comment (davthedev) [reply scene/resources/default_theme/default_theme.cpp:810]: Fair enough! I get it about the stability requirement between minor versions, and this totally makes sense. Godot gets more traction, game studios are moving to it and we want to avoid unpleasant surprises on their projects after an upgrade.

My vote is to have this change in. But we are a community and should decide together. So, go or no-go for this addition?

I have other open PRs with similar improvements on other types of controls. 
If we do not have a guideline about that acceptable scope of theme "breakage", I suggest we document it so that other contributors can be on the same wavelength.

Other options to avoid the surprise would be a good enough statement in the changelog, or detect that a project is from 4.0 and convert it at the moment it is opened.

Comment (YuriSizov) [reply scene/resources/default_theme/default_theme.cpp:810]: > So, go or no-go for this addition?

Keep it 🙃

**Review by YuriSizov — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!
